File: typescript-examples/captcha-solving-auth/api/list.ts
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";
import { extendPayload } from "@intuned/runtime";

interface Params {}

interface Product {
  name: string;
  price: string;
  details_url_item: string;
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<{ products: Product[] }> {
  /**
   * Scrapes products from a product grid page.
   * Extracts name, price, and detail URL for each product.
   *
   * Example params:
   * {
   *   "url": "https://scrapingcourse.com/dashboard"
   * }
   */

  await goToUrl({
    page,
    url: "https://scrapingcourse.com/dashboard",
  });

  const products: Product[] = [];
  const items = page.locator(".product-item");
  const count = await items.count();

  for (let i = 0; i < count; i++) {
    const item = items.nth(i);
    const name = (await item.locator(".product-name").textContent()) || "";
    const price = (await item.locator(".product-price").textContent()) || "";
    const details_url = (await item.locator("a").getAttribute("href")) || "";

    products.push({
      name: name.trim(),
      price: price.trim(),
      details_url_item: details_url,
    });
  }

  console.log(`Total products scraped: ${products.length}`);

  for (const product of products) {
    extendPayload({
      api: "details",
      parameters: product,
    });
  }

  return { products };
}

File: python-examples/captcha-solving-auth-example/api/list.py
from playwright.async_api import Page, BrowserContext
from intuned_browser import go_to_url
from typing import TypedDict, List
from intuned_runtime import extend_payload


class Params(TypedDict):
    pass

class Product(TypedDict):
    name: str
    price: str
    details_url_item: str


async def automation(
    page: Page,
    params: Params | None = None,
    context: BrowserContext | None = None,
    **_kwargs,
):
    """
    Scrapes products from a product grid page.
    Extracts name, price, and detail URL for each product.

    Example params:
    {
        "url": "https://scrapingcourse.com/dashboard"
    }
    """

    await go_to_url(page, "https://scrapingcourse.com/dashboard")
    products: List[Product] = []
    items = page.locator('.product-item')
    count = await items.count()

    for i in range(count):
        item = items.nth(i)
        name = await item.locator('.product-name').text_content()
        price = await item.locator('.product-price').text_content()
        details_url = await item.locator('a').get_attribute('href')

        products.append({
            'name': name.strip() if name else '',
            'price': price.strip() if price else '',
            'details_url_item': details_url or '',
        })

    print(f"Total products scraped: {len(products)}")

    for product in products:
        extend_payload({
            "api": "product-details",
            "parameters": dict(product),
        })

    return {"products": products}